### PR TITLE
Add support for BasedSession

### PR DESCRIPTION
**Pull request by jaraco**
Following on the [suggestion in requests](https://github.com/kennethreitz/requests/issues/2554#issuecomment-109341010), I've implemented the suggestion. Please let me know if there's anything more you'd like to see to get this accepted. I use this technique all the time and find the code copy/pasted in several places and I'd like to consolidate it somewhere.

Thanks.

### CODE REVIEW

**Comment by jaraco** [comment]
Damn. Failures in Python 2.7 suggest `requests.Session` is an old-style class.

**Comment by Lukasa** [comment]
`requests.Session` is not an old-style class.

**Comment by jaraco** [comment]
> requests.Session is not an old-style class.

Oh! Of course! It's the mix-in that's old style. My bad.

**Comment by sigmavirus24** [comment]
@jaraco could you explain the benefit of having the `Based` mix-in and the `BasedSession` class separate like you do? I'm not sure I see the benefit at the moment.

**Comment by jaraco** [comment]
@sigmavirus24 My intention there was to make the underlying functionality available to other requests.Session subclasses that other users may have in their libs/apps. I'm happy to fold that together or make it private at your preference.

**Comment by sigmavirus24** [comment]
I'm just having trouble thinking of a case where they wouldn't just subclass `BasedSession` if they're already sub-classing a session. I think I must be missing something.

File: docs/index.rst [hunk 1]
```diff
@@ -36,6 +36,7 @@ Full Documentation
     dumputils
     formdata
     exceptions
+    sessions
```
Comment (sigmavirus24) [inline docs/index.rst]: I think you forgot to add this file too.

**Comment by jaraco** [comment]
I've updated the code to simplify the implementation to just one subclass, and added the missing docs. Anything else to consider before acceptance?

**Comment by sigmavirus24** [comment]
@jaraco just time. I'll probably get around to this tomorrow and add some more docs to it when I merge it.

Thanks! 🎉

**Comment by jaraco** [comment]
I just went to use this today and found to my chagrin it still isn't available. Anything I can do to speed the adoption?

**Comment by sigmavirus24** [comment]
Nope. This reminder is good enough. I'm going to put some time into this project tonight and cut a release. I'm so sorry I lost track of this.

**Comment by jaraco** [comment]
Thank you.

**Comment by sigmavirus24** [comment]
0.7.0 is up on PyPI with this in it.

**Comment by foxx** [comment]
Thanks @jaraco and @sigmavirus24 for getting this done, I completely forgot to send a PR!

**Pull request merged**